# Trabajo Práctico 1 de Referenciación en Geofísica (2026)

Luciano P. O. Mendoza (lmendoza@fcaglp.unlp.edu.ar)

## Notas acerca de Jupyter

- Para **ejecutar** una celda hay que seleccionarla y oprimir SHIFT+ENTER.
- Para agregar una nueva celda abajo oprimir B o usar el botón + del menu superior.
- Para agregar una nueva celda arriba oprimir A.
- Para **guardar** el cuaderno oprimir SHIFT+S o usar el botón *floppy disk* del menú superior.
- Usar el menú desplegable superior para seleccionar entre los tres tipos de celdas posibles
  - *Code* para calculo (por ejemplo, como calculadora científica)
  - *Markdown* para texto con formato, ecuaciones matemáticas, imágenes, etc.
  - *Raw* para texto simple.

# Ejercicios

## Ejercicio 1

### Enunciado

Dados la siguiente matriz $A$ y el vector columna $b$
$$
\begin{aligned}
A &=
\begin{pmatrix}
5 & 5 & 7 & 8\\
4 & 7 & 1 & 1\\
4 & 4 & 2 & 1\\
2 & 4 & 0 & 3
\end{pmatrix}\\
b &=
\begin{pmatrix}
2\\3\\7\\9
\end{pmatrix}
\end{aligned}
$$
calcule las siguientes operaciones
$$
B =A^{-1}\qquad
C = A^T\qquad
a = Ab\qquad
c = A^T b= C b\qquad
d = \mathrm{traza}(A)
$$

In [1]:
# resolver acá
1 + 1

2

### Resolución

In [2]:
# primero que nada, importamos el módulo numpy, empleando el "diminutivo" np...
import numpy as np

# e importamos el módulo de algebra lineal linalg...
from scipy import linalg

In [3]:
# definimos la matriz A (por filas)...
A = np.array([[5,5,7,8],
              [4,7,1,1],
              [4,4,2,1],
              [2,4,0,3]])

# definimos el vector columna b (por filas)...
b = np.array([[2],
              [3],
              [7],
              [9]])

# mostramos A y b...
print(A)
print(b)

[[5 5 7 8]
 [4 7 1 1]
 [4 4 2 1]
 [2 4 0 3]]
[[2]
 [3]
 [7]
 [9]]


Ahora, con los datos cargados en la memoria (la RAM) de la computadora, podemos realizar las operaciones solicitadas inmediatamente

In [4]:
B = linalg.inv(A) # esto debe ser una matriz cuadrada
C = A.transpose() # aplicamos el método de transponer
a = A @ b         # este debe ser un vector columna
c = C @ b         # este debe ser un vector columna
d = A.trace()     # esto resultará un escalar

print(B)
print(C)
print(a)
print(c)
print(d)

[[-0.16049383 -0.55555556  0.83950617  0.33333333]
 [ 0.0617284   0.44444444 -0.4382716  -0.16666667]
 [ 0.18518519  0.33333333 -0.31481481 -0.5       ]
 [ 0.02469136 -0.22222222  0.02469136  0.33333333]]
[[5 4 4 2]
 [5 7 4 4]
 [7 1 2 0]
 [8 1 1 3]]
[[146]
 [ 45]
 [ 43]
 [ 43]]
[[68]
 [95]
 [31]
 [53]]
17


## Ejercicio 2

### Enunciado

Dados los siguientes dos vectores de $\mathbb{R}^2$
$$
\begin{aligned}
\mathbf{a}&=
\begin{pmatrix}
2\\
3
\end{pmatrix}\\
\mathbf{b}&=
\begin{pmatrix}
-1\\
\phantom{-}4
\end{pmatrix}
\end{aligned}
$$
indicar si son o no ortogonales.

### Notas

* Utilice la función [sum](https://numpy.org/doc/stable/reference/generated/numpy.sum.html) de NumPy, el producto __elemento a elemento__ (o sea, *, no @), y la estructura [if](https://docs.python.org/3/tutorial/controlflow.html), de Python.
* Suponer aquí que Python trabaja con precisión infinita (algo incorrecto) y que, de ser ortogonales, el producto punto será **exactamente** cero. ¡En el siguiente ejercicio habrá que tener más cuidado!

In [5]:
# resolver acá
1 + 1

2

### Resolución

In [6]:
a = np.array([[2],
              [3]])
     
b = np.array([[-1],
              [ 4]])

p = np.sum(a * b)

# un método
if p == 0:
    print("¡Son ortogonales!")
else:
    print("¡No son ortogonales!")

# otro método
if p != 0:
    print("¡No son ortogonales!")
else:
    print("¡Son ortogonales!")

¡No son ortogonales!
¡No son ortogonales!


## Ejercicio 3

### Enunciado

Dados dos vectores $\mathbf{a},\mathbf{b}$ cualesquiera, pertenecientes a $\mathbb{R}^n$, escribir funciones que indiquen si son paralelos/antiparalelos, ortogonales o ninguno de las dos anteriores.

### Notas

* Tenner en cuenta que Python, por defecto, trabaja con precisión finita, por lo que comparaciones **exactas** de resultados no son seguras (por ejemplo, $=0$ o $\ne 0$ o $= 1$ o $\ne 1$). Es recomendable entonces comparar contra un **número muy pero muy pequeño**, utilizando la función [abs](https://numpy.org/doc/stable/reference/generated/numpy.absolute.html).
* En Python, la _norma 2_ (i.e., $|\quad|$) de un vector o matriz se calcula con las funciones de Numpy/SciPy [linalg.norm](https://numpy.org/doc/stable/reference/generated/numpy.linalg.norm.html).
* Definir subprogramas, subrutinas o funciones, en Python o en cualquier lenguaje, ayuda a hacer el código más claro, menos propenso a errores y más fácil de corregir o mejorar.
* ¡Recuerde que una función puede *llamar* o utilizar otras funciones ya definidas!
* ¡Es **muy importante** nunca nombrar variables o funciones utilizando nombres de constantes o funciones de Python! Está es una muy común fuente de error. Por ejemplo, nunca definir constantes que se llamen sin, cos, min, max, pi, e, tan, atan, etc.
* Recordar que existen constantes especiales "True" y "False" para indicar el resultado de las comparaciones. Por ejemplo

In [7]:
1 > 0

True

In [8]:
1 < 0

False

In [9]:
2 + 2 == 4

True

In [10]:
2 + 2 != 4

False

In [11]:
1.5 > -6

True

In [12]:
# resolver acá
1 + 1

2

### Resolución

Primero, definiremos una **función** que indique simplemente si dos vectores son ortogonales o no. Para eso utilizamos el *producto punto* o *producto interno* entre dos vectores de $\mathbb{R}^n$
$$
r = \mathbf{a}\cdot \mathbf{b} = \sum_{i=1}^n a_i b_i
$$
que indicará que $\mathbf{a}$ y $\mathbf{b}$ son ortogonales si y solo si $r=0$.

Pero, como por defecto Python representa los números reales con precisión finita (como la mayoría de los lenguajes: Fortran, Octave, C, etc.) debemos reemplazar está condición por $|r| \le$ *un número muy pequeño*.

Como esto es tan común en programación Python ya tiene definido ese *número pequeño*, se llama *eps* (este número puede variar de computadora a computadora...). Python, además, define automáticamente otras constantes útiles

In [13]:
np.finfo(float).eps

2.220446049250313e-16

In [14]:
np.e

2.718281828459045

In [15]:
np.pi

3.141592653589793

Entonces, definimos nuestra primer función en Python

In [16]:
# Función para comprobar si dos vectores son o no ortogonales.
# si r = 1 entonces SÍ son ortogonales
# si r = 0 entonces NO son ortogonales
# utilizamos el producto punto
# utilizamos la "precisión" de la computadora "eps"
def sonortogonales(a,b):
    p = np.sum(a*b)
    return np.abs(p) <= np.finfo(float).eps # True o False...    

Podemos probar inmediatamente la función ¡notando que es independiente de la dimensión del espacio donde viven los vectores!

In [17]:
# estos sí son ortogonales
sonortogonales(np.array([1,0]),np.array([0,1]))

True

In [18]:
# estos no son ortogonales
sonortogonales(np.array([1,0]),np.array([1,1]))

False

In [19]:
# estos son "casi" ortogonales...
sonortogonales(np.array([1,0,0,0,0,0]),np.array([1,0,0,0,0,0.000000001]))

False

In [20]:
# estos no son ortogonales
sonortogonales(np.array([1,0,0,0,0,0]),np.array([1,0,0,0,0,1]))

False

In [21]:
# el vector "nulo" es ortogonal a todos los demás...
sonortogonales(np.array([1,2,3]),np.array([0,0,0]))

True

De manera similar podemos ahora escribir una función que indique si dos vectores son o no son paralelos (o antiparalelos) usando el coseno del ángulo $\theta$ *subtendido* por los vectores
$$
\cos\theta = \frac{\mathbf{a}\cdot \mathbf{b}}{|\mathbf{a}||\mathbf{b}|}
$$
Donde $\cos\theta = 1$ indica paralelismo y $\cos\theta = -1$ indica antiparalelismo. Estás dos condiciones pueden *representarse* juntas como $|\cos\theta| - 1 = 0$.

In [22]:
# Función para comprobar si dos vectores son o no paralelos o antiparalelos.
# si r = 1 entonces SÍ son paralelos o antiparalelos
# si r = 0 entonces NO son paralelos ni antiparalelos
# utilizamos el producto punto y el ángulo subtendido
# utilizamos la "precisión" de la computadora "eps"
def sonparalelos(a,b):
    p = np.sum(a*b)
    costheta = p/(linalg.norm(a)*linalg.norm(b))
    return np.abs(np.abs(costheta) - 1) <= np.finfo(float).eps

In [23]:
# estos sí son paralelos
sonparalelos(np.array([1,1]),np.array([2,2]))

True

In [24]:
# estos sí son paralelos
sonparalelos(np.array([0,0,1,0]),44*np.array([0,0,1,0]))

True

In [25]:
# estos sí son antiparalelos
sonparalelos(np.array([1,0]),np.array([-2,0]))

True

In [26]:
# esto no son paralelos ni antiparalelos
sonparalelos(np.array([1,0,0]),np.array([0,1,1]))

False

In [27]:
# el vector nulo no tiene definido un paralelismo
# y nuestra función fallará...
sonparalelos(np.array([1,0]),np.array([0,0]))

/tmp/ipykernel_126604/2923419375.py:8: RuntimeWarning: invalid value encountered in scalar divide
  costheta = p/(linalg.norm(a)*linalg.norm(b))


False

Ahora podemos escribir una última función que de respuesta al enunciado del ejercicio. En este caso, no queremos que regrese ningún resultado, sino que **escriba** alguna de las tres respuestas: *ortogonales*, *paralelos/antiparalelos* o *ni ortogonales ni paralelos/antiparalelos*. ¡Notar que nuestra función **utilizará** las dos funciones anteriores!

In [28]:
# función que escribe la respuesta correcta
def f(a,b):
    if sonortogonales(a,b):
        print("ortogonales")
        return
    
    if sonparalelos(a,b):
        print("paralelos/antiparalelos")
        return
    
    print("ni ortogonales ni paralelos/antiparalelos")

Ahora probamos la función *f*

In [29]:
# definimos varios vectores de prueba
a = np.array([1,2])
b = -3*a
c = np.array([-2,1])
d = np.array([1,3])

In [30]:
# comprobemos que a y b son antiparalelos
f(a,b)

paralelos/antiparalelos


In [31]:
# comprobemos que a y c son ortogonales
f(a,c)

ortogonales


In [32]:
# comprobemos que b y c son ortogonales
f(b,c)

ortogonales


In [33]:
# comprobemos que c y d no son ni ortogonales ni paralelos
f(c,d)

ni ortogonales ni paralelos/antiparalelos


## Ejercicio 4

### Enunciado

El archivo [sismos.txt](sismos.txt) contiene la época (año, mes, día, hora, minuto y
segundo), la posición del hipocentro (latitud, longitud y profundidad) y la magnitud de
sismos ocurridos en todo el mundo.

De este archivo extraer, utilizando Python, la siguiente información:

1. Cuándo y donde se produjo el terremoto de mayor magnitud.
2. Cuando y donde se produjeron los terremotos de mayor y menor profundidad.
3. El año en qué se produjeron la mayor cantidad de sismos en el mundo.
4. Filtrar el archivo original y generar una tabla (i.e., una matriz) con los terremotos en Sudamérica (aproximar los límites del continente con un rectángulo, buscando los vértices en una mapa o con el _software_ [GoogleEarth](https://earth.google.com/web/)).
5. Buscar el terremoto de mayor magnitud ocurrido en Sudamérica, y decir cuándo y dónde se produjo.
6. Salvar en un archivo de texto la tabla con los sismos en Sudamérica obtenida en el inciso 4.

### Notas

* Python permite *leer* (es decir, cargar en la memoria RAM) archivos de datos, tanto en formato texto como en formato binario. Así mismo, Python permite *salvar* (es decir, guardar) resultados o datos en archivos, tanto de texto como binarios.
* Pueden utilizarse diversas funciones para leer o escribir acrhivos de datos.
* La función más sencilla para *cargar* datos, aunque algo limitada, es [loadtxt](https://numpy.org/doc/stable/reference/generated/numpy.loadtxt.html).
* La función más sencilla para *guardar* el valor de las variables o los resultados es [savetxt](https://numpy.org/doc/stable/reference/generated/numpy.savetxt.html).

In [34]:
# resolver acá
1 + 1

2

### Resolución

Primero, debemos *leer* las columnas del archivo [sismos.txt](sismos.txt).

**Siempre** tenemos que abrir el archivo de datos con un visor de texto **antes** de cargarlo en Python, y constatar visualmente los datos, mirando, por ejemplo
* ¿de qué tipo son (enteros, reales, texto)?
* ¿En qué unidades están (metros, kilómetros, grados, radianes, etc.)?
* ¿Hay coordenadas geográficas, o geodésicas, o cartesianas, o proyectadas, etc.?
* ¿Tiene *agujeros* (renglones en blanco)?
* ¿Todas las columnas tienen la misma extensión? Es decir, ¿todas las filas de datos están *completas*?) 

Entonces, podemos cargar el archivo así

In [35]:
A = np.loadtxt("sismos.txt")

Lo primero que podemos hacer es averiguar cuantos sismos componenen la lista

In [36]:
nsismos = np.shape(A)[0] # cuantas filas tiene A?
print(nsismos)

944


Ahora podemos ver, por ejemplo, los datos del primer sismo listado

In [37]:
i = 0
print(A[i,:])

[ 2.002e+03  1.000e+00  1.000e+00  1.000e+01  3.900e+01  6.820e+00
 -5.521e+01 -1.290e+02  1.000e+01  6.000e+00]


O podemos ver, por ejemplo, los datos de los sismos 100 al 102 (tres sismos)

In [38]:
i = range(99,102)
print(A[i,:])

[[ 2.0020e+03  1.0000e+01  8.0000e+00  1.0000e+00  2.3000e+01  5.8800e+01
  -4.2040e+01  8.8380e+01  1.0000e+01  6.0000e+00]
 [ 2.0020e+03  1.0000e+01  1.0000e+01  1.0000e+01  5.0000e+01  2.0570e+01
  -1.7600e+00  1.3430e+02  1.0000e+01  7.6000e+00]
 [ 2.0020e+03  1.0000e+01  1.0000e+01  1.2000e+01  2.8000e+01  2.5800e+01
  -1.5100e+00  1.3397e+02  1.0000e+01  6.7000e+00]]


O podemos ver, por ejemplo, los datos del  último sismo listado

In [39]:
i = nsismos - 1 # los índice empiezan en 0!!!
print(A[i,:])

[2007.     12.     26.     22.      4.     56.19   52.58 -168.2    35.4
    6.4 ]


Ahora estamos listos para resolver inciso a inciso.

##### Inciso 1

In [40]:
# sismo de mayor magnitud
i = 9 # índice de la columan de magnitudes...
magmax = np.max(A[:,i])
imagmax = np.argmax(A[:,i])

print(magmax,imagmax)

9.0 435


Esto nos indica que el sismo de máxima magnitud está listado en la fila 436 (¡los índices comienzan en 0!), y que su magnitud fue 9.

Con esta información podemos inmediatamente responder al donde y el cuando

In [41]:
j = imagmax
print(A[j,:])

[2004.     12.     26.      0.     58.     53.45    3.29   95.98   30.
    9.  ]


#### Inciso 2

In [42]:
# sismo a mayor profundidad
i = 8 # índice de la columna de profundidades...
profmax = np.max(A[:,i])
iprofmax = np.argmax(A[:,i])
print(profmax,iprofmax)

j = iprofmax
print(A[j,:])

688.0 827
[ 2.0070e+03  5.0000e+00  6.0000e+00  2.2000e+01  1.0000e+00  8.9200e+00
 -1.9410e+01 -1.7931e+02  6.8800e+02  6.1000e+00]


In [43]:
# sismo a menor profundidad
i = 8 # índice de la columna de profundidades...
profmin = np.min(A[:,i])
iprofmin = np.argmin(A[:,i])
print(profmin,iprofmin)

j = iprofmin
print(A[j,:])

0.0 317
[ 2.004e+03  2.000e+00  2.400e+01  2.000e+00  2.700e+01  4.623e+01
  3.514e+01 -4.000e+00  0.000e+00  6.400e+00]


#### Inciso 3

Primero debemos obtener una lista de todos los años incluidos en el archivo, con la función *unique*

In [44]:
i = 0 # índice de la columna de años...
[ListaAnios,CantidadPorAnio] = np.unique(A[:,i], return_counts = True)
print(ListaAnios)
print(CantidadPorAnio)

[2002. 2003. 2004. 2005. 2006. 2007.]
[139 156 156 150 153 190]


Donde vemos que hay datos de seis años (de 2002 a 2007, inlcuido), y obtenemos el número de sismos (es decir, filas) a lo largo de los años.

Por ejemplo, ahora sabemos que en 2002 ocurrieron 139 sismos.

Ahora ya podemos resolver el inciso

In [45]:
# en que año ocurrieron más sismos
CantidadMax = np.max(CantidadPorAnio)
iCantidadMax = np.argmax(CantidadPorAnio)

AnioConMas = ListaAnios[iCantidadMax]
print(CantidadMax,iCantidadMax,AnioConMas) # los índices arrancan en cero!!!

190 5 2007.0


#### Inciso 4

Primero debemos definir los límites Oeste/Este y Sur/Norte de Sudamérica, buscando en un mapa, Wikipedia o Googleearth. Por ejemplo, podrían ser así

In [46]:
# límites de Sudamérica
lonmin = -82.5 # grados
lonmax = -33.0 # grados
latmin = -56.5 # grados
latmax = +13.0 # grados
print(lonmin,lonmax,latmin,latmax)

-82.5 -33.0 -56.5 13.0


Ahora **buscamos o filtramos** todos los sismos (es decir, todas las filas) que cumplan **sumultáneamente** las siguientes condiciones
$$
\lambda_\textrm{min} \le \lambda \le \lambda_\textrm{max}\qquad\varphi_\textrm{min} \le \varphi \le \varphi_\textrm{max}
$$

In [47]:
# ¿qué filas satisfacen simultáneamente las CUATRO condiciones?
# acá usamos que True*True = True, pero True*False = False y False*False = False
# es decir, & es el AND lógico, pero aplicado elemento a elemento...
ilat = 6 # índice de la columna de latitudes...
ilon = 7 # índice de la columna de longitudes...
i = (lonmin <= A[:,ilon]) & (lonmax >= A[:,ilon]) & (latmin <= A[:,ilat]) & (latmax >= A[:,ilat])

# ¿cuantas filas cunplen las condiciones, es decir, son True?
# acá usamos que False equivale a 0, y True equivale a no cero...
print(np.count_nonzero(i))

# ¿que porcentaje del total son?
print(100*np.count_nonzero(i)/nsismos)

81
8.580508474576272


Ahora estamos en condiciones de definir un nuevo conjunto de datos, que solo contenga las filas que encontramos

In [48]:
AenSA = A[i,:]

#### Inciso 5

In [49]:
# sismo de mayor magnitud en Sudamérica
i = 9 # índice de la columna de magnitudes...
magmax = np.max(AenSA[:,i])
imagmax = np.argmax(AenSA[:,i])

print(AenSA[imagmax,:])

[2007.      8.     15.     23.     40.     57.89  -13.39  -76.6    39.
    8.  ]


#### Inciso 6

Ahora podemos *salvar* (es decir, guardar) el resultado del filtrado o búsqueda de datos en Sudamérica en un archivo de texto.

In [50]:
np.savetxt("sismosenSA.txt",AenSA)

Naturalmente, **debemos** verificar el éxito de la operación abriendo el archivo *sismosenSA.txt* y verificando visualmente que es correcto.

### Resolución completa

Al fin, el ejercicio completo podría ser resuelto así

In [51]:
A = np.loadtxt("sismos.txt")

nsismos = np.shape(A)[0] # cuantas filas tiene A?
print(nsismos)

# inciso 1

# sismo de mayor magnitud
i = 9 # índice de la columan de magnitudes...
magmax = np.max(A[:,i])
imagmax = np.argmax(A[:,i])
print(magmax,imagmax)

j = imagmax
print(A[j,:])

# inciso 2

# sismo a mayor profundidad
i = 8 # índice de la columna de profundidades...
profmax = np.max(A[:,i])
iprofmax = np.argmax(A[:,i])
print(profmax,iprofmax)

j = iprofmax
print(A[j,:])

# sismo a menor profundidad
i = 8 # índice de la columna de profundidades...
profmin = np.min(A[:,i])
iprofmin = np.argmin(A[:,i])
print(profmin,iprofmin)

j = iprofmin
print(A[j,:])

# inciso 3

i = 0 # índice de la columna de años...
[ListaAnios,CantidadPorAnio] = np.unique(A[:,i], return_counts = True)
print(ListaAnios)
print(CantidadPorAnio)

# en que año ocurrieron más sismos
CantidadMax = np.max(CantidadPorAnio)
iCantidadMax = np.argmax(CantidadPorAnio)

AnioConMas = ListaAnios[iCantidadMax]
print(CantidadMax,iCantidadMax,AnioConMas) # los índices arrancan en cero!!!

# inciso 4

# límites de Sudamérica
lonmin = -82.5 # grados
lonmax = -33.0 # grados
latmin = -56.5 # grados
latmax = +13.0 # grados
print(lonmin,lonmax,latmin,latmax)

# ¿qué filas satisfacen simultáneamente las CUATRO condiciones?
# acá usamos que True*True = True, pero True*False = False y False*False = False
# es decir, & es el AND lógico, pero aplicado elemento a elemento...
ilat = 6 # índice de la columna de latitudes...
ilon = 7 # índice de la columna de longitudes...
i = (lonmin <= A[:,ilon]) & (lonmax >= A[:,ilon]) & (latmin <= A[:,ilat]) & (latmax >= A[:,ilat])

# ¿cuantas filas cunplen las condiciones, es decir, son True?
# acá usamos que False equivale a 0, y True equivale a no cero...
print(np.count_nonzero(i))

# ¿que porcentaje del total son?
print(100*np.count_nonzero(i)/nsismos)

AenSA = A[i,:]

# inciso 5

# sismo de mayor magnitud en Sudamérica
i = 9 # índice de la columna de magnitudes...
magmax = np.max(AenSA[:,i])
imagmax = np.argmax(AenSA[:,i])

print(AenSA[imagmax,:])

# inciso 6

np.savetxt("sismosenSA.txt",AenSA)
print("¡Datos salvados!")

944
9.0 435
[2004.     12.     26.      0.     58.     53.45    3.29   95.98   30.
    9.  ]
688.0 827
[ 2.0070e+03  5.0000e+00  6.0000e+00  2.2000e+01  1.0000e+00  8.9200e+00
 -1.9410e+01 -1.7931e+02  6.8800e+02  6.1000e+00]
0.0 317
[ 2.004e+03  2.000e+00  2.400e+01  2.000e+00  2.700e+01  4.623e+01
  3.514e+01 -4.000e+00  0.000e+00  6.400e+00]
[2002. 2003. 2004. 2005. 2006. 2007.]
[139 156 156 150 153 190]
190 5 2007.0
-82.5 -33.0 -56.5 13.0
81
8.580508474576272
[2007.      8.     15.     23.     40.     57.89  -13.39  -76.6    39.
    8.  ]
¡Datos salvados!
